File: model/src/main/java/io/github/chriszhong/model/organization/event/CharacteristicEvent.java
package io.github.chriszhong.model.organization.event;

import io.github.chriszhong.message.M;
import io.github.runtimemodels.chazm.entity.Characteristic;
import io.github.runtimemodels.chazm.id.UniqueId;

import java.util.Objects;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;

import com.google.inject.assistedinject.Assisted;

/**
 * The {@linkplain CharacteristicEvent} class indicates that there is an update about a {@linkplain Characteristic} entity.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public class CharacteristicEvent extends AbstractEvent {

	private static final long serialVersionUID = -7464874692386039815L;
	private final UniqueId<Characteristic> id;
	private transient Integer hashCode = null;
	private transient String toString = null;

	@Inject
	CharacteristicEvent(@NotNull @Assisted final EventCategory category, @NotNull @Assisted final Characteristic characteristic) {
		super(category);
		id = characteristic.getId();
	}

	/**
	 * Returns a {@linkplain UniqueId} that represents a {@linkplain Characteristic}.
	 *
	 * @return a {@linkplain UniqueId}.
	 */
	public UniqueId<Characteristic> getId() {
		return id;
	}

	@Override
	public boolean equals(final Object object) {
		if (object instanceof CharacteristicEvent) {
			final CharacteristicEvent event = (CharacteristicEvent) object;
			return super.equals(event) && getId().equals(event.getId());
		}
		return false;
	}

	@Override
	public int hashCode() {
		if (hashCode == null) {
			hashCode = Objects.hash(getCategory(), getId());
		}
		return hashCode;
	}

	@Override
	public String toString() {
		if (toString == null) {
			toString = M.EVENT_WITH_1_ID.get(super.toString(), getId());
		}
		return toString;
	}

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/entity/CharacteristicEvent.kt
package runtimemodels.chazm.model.entity

import com.google.inject.assistedinject.Assisted
import runtimemodels.chazm.api.entity.Characteristic
import runtimemodels.chazm.api.entity.CharacteristicId
import runtimemodels.chazm.api.id.UniqueId
import runtimemodels.chazm.model.event.AbstractEvent
import runtimemodels.chazm.model.event.EventType
import runtimemodels.chazm.model.message.M
import java.util.*
import javax.inject.Inject

/**
 * The [CharacteristicEvent] class indicates that there is an update about a [Characteristic] entity.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
open class CharacteristicEvent @Inject internal constructor(
    @Assisted category: EventType,
    @Assisted characteristic: Characteristic
) : AbstractEvent(category) {
    /**
     * Returns a [UniqueId] that represents a [Characteristic].
     *
     * @return a [UniqueId].
     */
    val id: CharacteristicId = characteristic.id
    @Transient
    private var hashCode: Int? = null
    @Transient
    private var toString: String? = null

    override fun equals(other: Any?): Boolean {
        if (other is CharacteristicEvent) {
            return super.equals(other) && id == other.id
        }
        return false
    }

    override fun hashCode(): Int = Objects.hash(category, id)

    override fun toString(): String = M.EVENT_WITH_1_ID[super.toString(), id]

    companion object {
        private const val serialVersionUID = -7464874692386039815L
    }

}
